File: assn3/qsort.py
#!/usr/bin/python3

def partition(array, low, high):
    index = low-1
    pivot = array[high]

    for i in range(low, high):
        if array[i] <= pivot:
            index = index + 1
            array[index], array[i] = array[i], array[index]
    
    array[index+1], array[high] = array[high], array[index+1]
    return(index+1)

def qsort(array, low, high):
    if low < high:
        newIndex = partition(array,low,high)
        qsort(array, low, newIndex-1)
        qsort(array, newIndex+1, high)

def main():
    arrSize = input("Enter Size of Array: ")
    A = [0]*int(arrSize)

    for i in range(0,int(arrSize)):
        arrNums = input("Enter the value for position A[%d]:\n" % (i))
        A[i] = int(arrNums)

    print("Input:", A)
    qsort(A,0,len(A)-1)
    print("Output:", A, "\n\n")

main()

File: assn3/qsort.cpp
#include <iostream>
#include <vector>

using namespace std;

void swap(int* a, int* b){
    int t = *a;
    *a = *b;
    *b = t;
}

int partition(int array[], int low, int high){
    int pivot = array[high];
    int index = low-1;
    int loop = high - low;

    for(int i=low; i<=high-1; i++){
        if(array[i] <= pivot){
            index++;
            swap(&array[index], &array[i]);
        }
    }
    swap(&array[index+1], &array[high]);
    return index+1;
}

void qsort(int array[], int low, int high){
    if(low<high){
        int newIndex = partition(array,low,high);
        qsort(array, low, newIndex-1);
        qsort(array, newIndex+1, high);
    }
}

int main() {
      int arrSize;
      cout << "Enter size of Array:" << endl;
      cin >> arrSize;

      int A[arrSize];

      for(int i=0; i<arrSize; i++){
            cout << "Enter value for position A[" << i << "]" << endl;
            cin >> A[i];
      }

      cout << "Input: \n[";
      for(int k=0; k<arrSize; k++){
            if(k == arrSize-1){
                  cout << A[k];
            } else {
                  cout << A[k] << ", ";
            }
      }

      qsort(A, 0, arrSize-1);

      cout << "]\nOutput: \n[";
      for(int k=0; k<arrSize; k++){
            if(k == arrSize-1){
                  cout << A[k];
            } else {
                  cout << A[k] << ", ";
            }
      }
      cout << "]\n\n";

        return 0;
    }

